File: DSA Practice/LeetCode/938. Range Sum of BST/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rec(TreeNode root, int low, int high){
        if (root == null){
            return 0;
        }

        if (root.val >= low && root.val <= high){
            if (root.val == low){
                return root.val + rec(root.right, low, high);
            }

            else if (root.val == high){
                return root.val + rec(root.left, low, high);
            }

            else {
                return root.val + rec(root.left, low, high) + rec(root.right, low, high);
            }
        }

        else if (root.val > high){
            return rec(root.left, low, high);
        } 

        else if (root.val < low){
            return rec(root.right, low, high);
        }

        return 0;
    }

    public int rangeSumBST(TreeNode root, int low, int high) {
        int ans = rec(root, low, high);
        return ans;
    }
}

File: DSA Practice/LeetCode/938. Range Sum of BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:

        def rec(node):
            if node == None:
                return 0

            if high >= node.val >= low:
                if node.val == low:
                    return node.val + rec(node.right)
                elif node.val == high:
                    return node.val + rec(node.left)
                else:
                    return node.val + rec(node.left) + rec(node.right)

            elif node.val > high:
                return rec(node.left)

            elif node.val < low:
                return rec(node.right)

        ans = rec(root)

        print(ans)
        return ans